web/modules/custom/joinup_core/src/NodeWorkflowAccessControlHandler.php

```diff
@@ -454,21 +471,49 @@ protected function getParentElibraryName(EntityInterface $entity) {
   /**
    * Checks whether the entity has a published version.
    *
-   * @param \Drupal\Core\Entity\EntityInterface $entity
+   * @param \Drupal\node\NodeInterface $entity
    *   The entity object.
    *
    * @return bool
    *   Whether the entity has a published version.
    */
-  protected function hasPublishedVersion(EntityInterface $entity) {
+  protected function hasPublishedVersion(NodeInterface $entity): bool {
     if ($entity->isNew()) {
       return FALSE;
     }
     if ($entity->isPublished()) {
       return TRUE;
     }
-    $published = $this->entityTypeManager->getStorage('node')->load($entity->id());
-    return $published->isPublished();
+    $published = $this->getNodeStorage()->load($entity->id());
+    if (!empty($published) && $published instanceof EntityPublishedInterface) {
+      return $published->isPublished();
+    }
+    return FALSE;
+  }
+
+  /**
+   * Returns the configured permission scheme for the given operation.
+   *
+   * @param string $operation
+   *   The operation for which to return the permission scheme. Can be one of
+   *   'create', 'view', 'update', 'delete'.
+   *
+   * @return array
+   *   The permission scheme.
+   */
+  protected function getPermissionScheme(string $operation): array {
+    \assert(\in_array($operation, ['create', 'view', 'update', 'delete']), 'A valid operation should be passed');
+    return $this->configFactory->get('joinup_community_content.permission_scheme')->get($operation);
+  }
+
+  /**
+   * Returns the storage handler for nodes.
+   *
+   * @return \Drupal\node\NodeStorageInterface
+   *   The storage handler.
+   */
+  protected function getNodeStorage(): NodeStorageInterface {
```

Comment:
Yes this is just here to leverage strict typing, otherwise we need to add `/** @var */` or `instanceof` checks. I have been exploring other options but this is the easiest so far until we come up with a structural solution.

Other options are to make traits or a specific `NodeTypeManager` that overrides `EntityTypeManager` and provides the specific strict return types. I tried to discuss this with @sandervd but he was either too busy or was being tactfully silent instead of rejecting my proposal outright :)

It's indeed a bit academic but this simple method helps to satisfy my perfectionism a great deal.

Reply:
The answer satisfied me, too :)